File: hojong/uniquePaths.java
class Solution {
    
    public int uniquePaths(int m, int n) {
        int[][] map = new int[m][n];
        
        for (int row=0; row<m; row++) {
            for (int col=0; col<n; col++) {
                if (row == 0 || col == 0) {
                    map[row][col] = 1;
                } else {
                    map[row][col] = map[row][col-1] + map[row-1][col];
                }
            }
        } 
        
        return map[m-1][n-1];
    }
}

File: beomseok/UniquePaths.kt
class Solution {
    fun uniquePaths(m: Int, n: Int): Int {
        if (m == n && m == 1) {
            return 1
        }
        
        // 초기화
        // dp[i][0] dp[0][j] = 1
        val dp = Array<IntArray>(m) { _ -> IntArray(n) { _ -> 0 } }
        for (i in 0 until m) {
            for (j in 0 until n) {
                if (i == 0 || j == 0) {
                    dp[i][j] = 1
                }
            }
        }
        
        // dp 점화식 : dp[i][j] = dp[i-1][j] + dp[i][j-1]
        for (i in 1 until m) {
            for (j in 1 until n) {
                dp[i][j] = dp[i-1][j] + dp[i][j-1]
            }
        }
        
        return dp[m-1][n-1]
    }
}
